rfd/0031-dynamic-app-db-registration.md

```diff
@@ -0,0 +1,355 @@
+
+```yaml
+app_service:
+  enabled: "yes"
+  # Watch for apps with app=ops&env=dev or env=test labels.
+  selectors:
+  - matchLabels:
+      app: ops
+      env: dev
+  - matchLabels:
+      env: test
+  # Existing static apps configuration, may be omitted.
+  apps:
+    ...
+```
+
+```yaml
+db_service:
+  enabled: "yes"
+  # Watch for any database with env label set.
+  selector:
+  - matchLabels:
+      env: "*"
+  # Existing static databases configuration, may be omitted.
+  dbs:
+    ...
+```
+
+The `matchLabels` approach is inspired by Kubernetes selectors. It is the only
+type of selector supported initially, but gives us a way to introduce other
+types in future if necessary (e.g. K8s has `matchExpression`).
+
+To make a service watch for any app/database, an explicit wildcard selector
+must be specified:
+
+```yaml
+selector:
+- matchLabels:
+    "*": "*"
+```
+
+**Static vs dynamic configuration**
+
+There must be a way to distinguish between apps and databases from the static
+yaml configuration vs ones registered as dynamic resources. This is needed,
+for example, to prevent users from deleting those registered statically.
+
+We will use the same approach as in [RFD16 Dynamic Configuration](./0016-dynamic-configuration.md)
+and use `teleport.dev/origin` label to denote whether the app/database resource
+comes from static configuration or created dynamically. The rules are:
+
+- Users can't delete apps/databases with `teleport.dev/origin: config-file`
+  label.
+- Users can't set `teleport.dev/origin` label in their resources - it will be
+  auto-set to `dynamic` when they're created.
+- Users can't create apps/databases with the same name as one of existing
+  static apps/databases (to avoid various conflict resolution issues).
+
+**Multiple replicas**
+
+When multiple app/database agents pick up the same web app or database resource,
+they all add it to their proxying configuration. The connections that go to
+this app/database follow the same load-balancing rules which already exist
+today. This is no different than registering the same app/database multiple
+times via static configuration today.
+
+## Security implications
+
+In the current Teleport security model, the trust is established between
+components of the system i.e. when a new app or database service agent joins
+the cluster, it needs to present a valid auth token. Once the service has
+registered, updating its configuration to connect new web apps or databases
+does not require re-registering the service with the cluster.
+
+Dynamic app/database registration plugs into this model - when a new app or
+database resource is created, it is picked up by a service that is already
+part of the cluster. Managing app/database resources themselves is subject to
+RBAC checks so only users with appropriate permissions to `app_server` and
+`db_server` resources are allowed to create/remove them.
+
+## API changes
```

Comment:
Please describe any gRPC API changes here too.
Changing the Go API in a backwards-incompatible way is not ideal (for existing users), but doable.
Changing the gRPC API requires a lot more thought and can't be backwards-incompatible.